store/tikv/mocktikv/cop_handler_dag.go

```diff
@@ -370,6 +406,98 @@ func flagsToStatementContext(flags uint64) *stmtctx.StatementContext {
 	return sc
 }
 
+// mockClientStream implements grpc ClientStream interface, its methods are never called.
+type mockClientStream struct{}
+
+func (mockClientStream) Header() (metadata.MD, error) { return nil, nil }
+func (mockClientStream) Trailer() metadata.MD         { return nil }
+func (mockClientStream) CloseSend() error             { return nil }
+func (mockClientStream) Context() goctx.Context       { return nil }
+func (mockClientStream) SendMsg(m interface{}) error  { return nil }
+func (mockClientStream) RecvMsg(m interface{}) error  { return nil }
+
+type mockCopStreamClient struct {
+	mockClientStream
+
+	req      *tipb.DAGRequest
+	exec     executor
+	goCtx    goctx.Context
+	finished bool
+}
+
+type mockCopStreamErrClient struct {
+	mockClientStream
+
+	*errorpb.Error
+}
+
+func (mock *mockCopStreamErrClient) Recv() (*coprocessor.Response, error) {
+	return &coprocessor.Response{
+		RegionError: mock.Error,
+	}, nil
+}
+
+func (mock *mockCopStreamClient) Recv() (*coprocessor.Response, error) {
+	if mock.finished {
+		return nil, io.EOF
+	}
+
+	var resp coprocessor.Response
+	chunk, finish, err := mock.readBlockFromExecutor()
+	if err != nil {
+		if locked, ok := errors.Cause(err).(*ErrLocked); ok {
+			resp.Locked = &kvrpcpb.LockInfo{
+				Key:         locked.Key,
+				PrimaryLock: locked.Primary,
+				LockVersion: locked.StartTS,
+				LockTtl:     locked.TTL,
+			}
+		} else {
+			resp.OtherError = err.Error()
+		}
+		return &resp, nil
+	}
+	if finish {
+		// Just mark it, need to handle the last chunk.
+		mock.finished = true
+	}
+
+	data, err := chunk.Marshal()
+	if err != nil {
+		resp.OtherError = err.Error()
+		return &resp, nil
+	}
+	streamResponse := tipb.StreamResponse{
+		Error:      toPBError(err),
+		EncodeType: tipb.EncodeType_TypeDefault,
+		Data:       data,
+	}
+	data1, err1 := proto.Marshal(&streamResponse)
+	if err1 != nil {
+		resp.OtherError = err1.Error()
+		return &resp, nil
+	}
+	resp.Data = data1
+	return &resp, nil
+}
+
+func (mock *mockCopStreamClient) readBlockFromExecutor() (tipb.Chunk, bool, error) {
+	var chunk tipb.Chunk
+	for count := 0; count < rowsPerChunk; count++ {
+		row, err := mock.exec.Next(mock.goCtx)
+		if err != nil {
+			return chunk, false, errors.Trace(err)
```

Comment:
Returning nil is better?

Reply:
The return type is `tipb.Chunk`, not `*tipb.Chunk`, so I can't return nil here.
I guess it doesn't make any difference.